### PR TITLE
Reintroduce Phalanx special DynRankView operator[]

### PR DESCRIPTION
**Pull request by crtrott**
@ndellingwood @rppawlo 

I am reintroducing some special sauce for Phalanx which got lost in the rewrite despite the very explicit comments in the old code: https://github.com/kokkos/kokkos/blob/15dc143e5f39949eece972a798e175c4b463d4b8/containers/src/Kokkos_DynRankView.hpp#L585-L621

However, the comment also indicates that that was meant as more of a temporary accommodation not a thing we support in perpetua so this PR introduces this behavior back in ONLY with deprecated code enabled, and in debug mode will check that the operator[] is only called if the dynamic rank is 1.

@rppawlo if this is absolutely unacceptable we can talk about this. But I have fairly strong reservations about carrying this behavior forward. 

Hopefully fixes all of: https://github.com/kokkos/kokkos/issues/7375

### CODE REVIEW

**Comment by ndellingwood** [comment]
Thanks @crtrott ! Changes look reasonable, I'll follow up after testing

**Review by ndellingwood — APPROVED** [review]
The phalanx tests failing in https://github.com/kokkos/kokkos/issues/7375 are passing, thanks @crtrott ! Should wait on Roger's feedback before final approval/merge

**Comment by ndellingwood** [comment]
Huh, this also fixed the failing Panzer subtest from #7375 as well, even though that test didn't use Fad types...

**Comment by rppawlo** [comment]
Thanks for fixing @crtrott and @ndellingwood ! I'm all for deprecating this code. In the failing phalanx unit test, I found this comment I made in 2016 during the initial transition to kokkos:
```
    // In the initial implementation of the Kokkos::DynRankView                                                                                                                          
    // bracket operator, you could only use it on a rank-1 view. This                                                                                                                    
    // is not how Intrepid uses it. Intrepid expects access to the                                                                                                                       
    // entire size of the array. This is dangerous and will not work                                                                                                                     
    // with subviews/padding. So the bracket op was changed to support                                                                                                                   
    // this but is considered a hack until we can remove all bracket                                                                                                                     
    // op use.               
```

File: containers/src/Kokkos_DynRankView.hpp
Comment (masterleinad) [inline containers/src/Kokkos_DynRankView.hpp:610]: This looks like a static property so we don't need a runtime check
```suggestion
    static_assert(view_type::rank() == 1u, "DynRankView operator[] can only be used for rank-1");
```
but it's also defined as
```
  using view_type = View<DataType*******, Properties...>;
```
so it wouldn't trigger anyway. What am I missing?

Comment (crtrott) [reply containers/src/Kokkos_DynRankView.hpp:610]: yeah lol you are right it needs to call `rank()`

File: containers/src/Kokkos_DynRankView.hpp
Comment (nmm0) [inline containers/src/Kokkos_DynRankView.hpp:600]: Maybe a dumb question, but aren't these all booleans? Thus the memory traits would either be 0 or Unmanaged?

File: containers/src/Kokkos_DynRankView.hpp
Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp:610]:
```suggestion
#ifdef KOKKOS_ENABLE_DEBUG
    if (rank() != 1u)
      Kokkos::abort("DynRankView operator[] can only be used for rank-1");
#endif
```

File: containers/src/Kokkos_DynRankView.hpp
Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp:600]: No, you probably want `MemoryTraitsFlags`
```suggestion
          Kokkos::MemoryTraits<Kokkos::Unmanaged |
                               Kokkos::RandomAccess |
                               Kokkos::Atomic>>;
```

Comment (nmm0) [reply containers/src/Kokkos_DynRankView.hpp:600]: Would you want to take the values from traits, rather than enabling all three flags?

Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp:600]: Like
```
(traits::memory_traits::is_unmanaged ? Kokkos::Unmanaged : 0) | ...
```

Comment (nmm0) [reply containers/src/Kokkos_DynRankView.hpp:600]: yeah but I'm honestly not sure what this function is supposed to do

Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp:600]: To me it looks like we want to reinterpret the DynRankView as 1-D View to access all elements preserving the memory traits (ignoring `Restrict` and `Aligned`).

Comment (nmm0) [reply containers/src/Kokkos_DynRankView.hpp:600]: @crtrott just copied the old code over, so it looks like this might have been broken for the last 5 years? :P

Comment (masterleinad) [reply containers/src/Kokkos_DynRankView.hpp:600]: I mean it apparently worked good enough for `Trilinos`. 🙂

**Review by nmm0 — CHANGES_REQUESTED** [review]
I guess we should probably fix the memory traits bug

**Comment by crtrott** [comment]
Fixed both the Rank and the MemoryTraits thing.

**Review by masterleinad — APPROVED** [review]

**Review by ndellingwood — APPROVED** [review]

**Review by nmm0 — APPROVED** [review]
Looks good to me!

File: containers/src/Kokkos_DynRankView.hpp
Comment (dalg24) [inline containers/src/Kokkos_DynRankView.hpp:584]: "who uses" or "that is using"?

File: containers/src/Kokkos_DynRankView.hpp
Comment (dalg24) [inline containers/src/Kokkos_DynRankView.hpp:589]:
```suggestion
    if constexpr (std::is_same_v<typename drvtraits::value_type,
                               typename drvtraits::scalar_array_type>) {
```
(not formatted)

I thought we were enforcing this?!

File: containers/src/Kokkos_DynRankView.hpp
Comment (dalg24) [inline containers/src/Kokkos_DynRankView.hpp:599]: I can't believe we did not mandate the underlying integral type of `MemoryTraitsFlags` is not `unsigned` in the first place and we cast everywhere...

Comment (dalg24) [reply containers/src/Kokkos_DynRankView.hpp:599]: Yes sorry I am just ruminating.  Didn't mean to suggest there is anything to fix here

File: containers/src/Kokkos_DynRankView.hpp
Comment (dalg24) [inline containers/src/Kokkos_DynRankView.hpp:584]: Haha you misspelled usin~~t~~g
(I don't really care to fix)

**Pull request merged**
